src/caffe/layers/data_layer.cpp

```diff
@@ -14,7 +14,31 @@ namespace caffe {
 template <typename Dtype>
 DataLayer<Dtype>::DataLayer(const LayerParameter& param)
   : BasePrefetchingDataLayer<Dtype>(param),
-    reader_(param), pool_(3) {
+    reader_(param) {
+    if (this->layer_param_.data_param().threads() == 0) {
+#ifndef CPU_ONLY
+        if (Caffe::mode() == Caffe::GPU) {
+            int cuda_devices;
+            int cpu_count;
+            int prefetch_threads;
+            cpu_count = boost::thread::hardware_concurrency();
+            cudaGetDeviceCount(&cuda_devices);
+            DLOG(INFO) << "CPU threads: " << cpu_count << " GPU devices visible: " << cuda_devices;
+            
+            prefetch_threads = cpu_count > 1 ? cpu_count / cuda_devices : 1;
+            DLOG(INFO) << "Data processing threads: " << prefetch_threads;
+            pool_ = new ThreadPool(prefetch_threads);
+        }
+        else {
+            pool_ = new ThreadPool(1);
+        }
+#else
+        pool_ = new ThreadPool(1);
+#endif
+    }
+    else {
+        pool_ = new ThreadPool(this->layer_param_.data_param().threads());
```

Comment:
I don't see complimentary `delete` call in destructor. Not a problem if we use `shared_ptr<ThreadPool>`.